src/tools/kata-ctl/src/exec.rs

```diff
@@ -0,0 +1,337 @@
+    }
+
+    if msg.starts_with(OK_CMD) {
+        let response = msg
+            .strip_prefix(OK_CMD)
+            .ok_or(format!("invalid response: {:?}", msg))
+            .map_err(|e| anyhow!(e))?
+            .trim();
+        debug!(sl!(), "Hybrid Vsock host-side port: {:?}", response);
+        // Unset the timeout in order to turn the sokect to bloking mode.
+        stream.set_read_timeout(None)?;
+        stream.set_write_timeout(None)?;
+    } else {
+        stdin_handle
+            .lock()
+            .set_canon_mode()
+            .expect("Fail to set stdin to Canon mode in starts_with");
+
+        return Err(anyhow!(
+            "failed to setup Hybrid Vsock connection: response was: {:?}",
+            msg
+        ));
+    }
+
+    Ok(())
+}
+
+fn create_ttrpc_socket(
+    stdin_handle: io::Stdin,
+    server_addr: String,
+    hybrid_vsock_port: u64,
+) -> anyhow::Result<UnixStream> {
+    let stream = match UnixStream::connect(server_addr) {
+        Ok(s) => s,
+        Err(e) => {
+            return Err(anyhow!(e).context("failed to create named UNIX Domain stream socket"))
+        }
+    };
+
+    // when vport > 0, setup hvsock.
+    if hybrid_vsock_port > 0 {
+        setup_hybrid_vsock(stdin_handle, &stream, hybrid_vsock_port)?
+    }
+
+    Ok(stream)
+}
+
+fn debug_console_uds(sid: &str, sandbox_run_path: &str) -> PathBuf {
+    // we use pattern to get the full sandbox run path,
+    let sandbox_id = format!("{}{}", sid, "*");
+
+    // ${KATA_RUN_PATH}/root/{SANDBOX_ID}/kata.hvsock
+    // TODO: get vsock path for other scenario
+    let sandbox_path: String = [
+        sandbox_run_path,
+        sandbox_id.as_str(),
+        "root",
+        KATA_HYBRID_VSOCK,
+    ]
+    .join("/");
+
+    glob(sandbox_path.as_str())
+        .unwrap()
+        .filter_map(std::result::Result::ok)
+        .last()
+        .unwrap_or_else(|| PathBuf::from(sandbox_path.as_str()))
+}
+
+pub fn do_run_exec(sandbox_id: &str, debug_port: u64) -> anyhow::Result<()> {
+    let server_addr = debug_console_uds(sandbox_id, KATA_RUN_PATH)
+        .into_os_string()
+        .into_string()
+        .unwrap();
+
+    if !PathBuf::from(server_addr.clone()).exists() {
+        return Err(anyhow!("sandbox with run path {:?} not found", server_addr));
+    }
+
+    let mut epoll_context = EpollContext::new().expect("Fail to create epoll");
+    let stdin_handle = io::stdin();
+    epoll_context
+        .enable_stdin_event()
+        .expect("Fail to enable stdin");
+
+    let sock_addr: UnixStream = {
+        stdin_handle
+            .lock()
+            .set_raw_mode()
+            .expect("Fail to set stdin to RAW mode");
+
+        // by kata agent hvsock
+        create_ttrpc_socket(stdin_handle, server_addr, debug_port)?
+    };
+
+    epoll_context
+        .enable_debug_console_sock(sock_addr)
+        .expect("Fail to connect to the server");
+
+    epoll_context.process_handler().unwrap_or_default();
```

Comment:
Maybe we should do the error handling here.

Reply:
Yes, I will fix it.